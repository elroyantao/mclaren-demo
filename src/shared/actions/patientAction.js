import 'isomorphic-fetch'

export function fetchPatientsRequest() {
  return {
    type: 'FETCH_PATIENTS_REQUEST'
  }
}

export function fetchPatientsResponse(patients) {
  return {
    type: 'FETCH_PATIENTS_RESPONSE',
    patients
  }
}

export function fetchPatientsFailure() {
  return {
    type: 'FETCH_PATIENTS_FAILURE'
  }
}

export const fetchPatients = () => (dispatch) => {
  dispatch(fetchPatientsRequest())
  return fetch('/api/patients', { method: 'GET' })
    .then((response) => {
      if (!response.ok) throw Error(response.statusText)
      return response.json()
    })
    .then((response) => {
      dispatch(fetchPatientsResponse(response))
    })
    .catch(() => {
      dispatch(fetchPatientsFailure())
    })
}

export function fetchPatientsActivityRequest() {
  return {
    type: 'FETCH_PATIENTSACTIVITY_REQUEST'
  }
}

export function fetchPatientsActivityResponse(activities) {
  return {
    type: 'FETCH_PATIENTSACTIVITY_RESPONSE',
    activities
  }
}

export function fetchPatientsActivityFailure() {
  return {
    type: 'FETCH_PATIENTSACTIVITY_FAILURE'
  }
}

export const fetchPatientsActivity = () => (dispatch) => {
  dispatch(fetchPatientsActivityRequest())
  return fetch('/api/activity', { method: 'GET' })
    .then((response) => {
      if (!response.ok) throw Error(response.statusText)
      return response.json()
    })
    .then((response) => {
      dispatch(fetchPatientsActivityResponse(response))
    })
    .catch(() => {
      dispatch(fetchPatientsActivityFailure())
    })
}

export function setList(patients) {
  return {
    type: 'SET_LIST',
    patients
  }
}

export function generatePatientIntensity(patientActivity, activities) {
  const initialIntensity = {
    moderate: 0,
    vigorous: 0
  }
  return patientActivity.reduce((prev, activity) => {
    if (activities[activity.activity] === 'moderate') {
      return {
        ...prev,
        moderate: prev.moderate + activity.minutes
      }
    } else if (activities[activity.activity] === 'vigorous') {
      return {
        ...prev,
        vigorous: prev.vigorous + activity.minutes
      }
    }
    return prev
  }, initialIntensity)
}

export const generatePatientList = () => (dispatch, getState) => {
  const { patients, activities: activityDefinition, patientActivities } = getState()
  const patientList = patients.reduce((prev, patient) => {
    const activities = patientActivities[patient.id]
    const intensity = generatePatientIntensity(activities, activityDefinition)
    return [...prev, { ...patient, intensity, activities }]
  }, [])
  // const filteredList = applyFilters(list, filters)
  dispatch(setList(patientList))
}
