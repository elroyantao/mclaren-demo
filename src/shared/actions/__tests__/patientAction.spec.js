import fetchMock from 'fetch-mock'
import * as actions from '../patientAction'

const fetchAction = actions.fetchPatients()
const fetchPatientsActivity = actions.fetchPatientsActivity()
const dispatch = jest.fn()

describe('patient actions', () => {
  it('fetchPatientsRequest()', () => {
    expect(actions.fetchPatientsRequest()).toMatchSnapshot()
  })
  it('fetchPatientsFailure()', () => {
    expect(actions.fetchPatientsFailure()).toMatchSnapshot()
  })
  it('fetchPatientsResponse()', () => {
    expect(actions.fetchPatientsResponse([1, 2, 3])).toMatchSnapshot()
  })
  describe('fetchPatients() ', () => {
    // fetch.mockResponse(JSON.stringify({ access_token: '12345' }))
    afterEach(() => {
      fetchMock.restore()
      jest.clearAllMocks()
    })
    it('a successful fetch', () => {
      fetchMock.get('/api/patients', {
        data: [1, 2, 3]
      })
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/patients', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
    it('an unsuccessful fetch', () => {
      fetchMock.get('/api/patients', 402)
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/patients', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
  })

  it('fetchPatientsActivityRequest()', () => {
    expect(actions.fetchPatientsActivityRequest()).toMatchSnapshot()
  })
  it('fetchPatientsActivityFailure()', () => {
    expect(actions.fetchPatientsActivityFailure()).toMatchSnapshot()
  })
  it('fetchPatientsActivityResponse()', () => {
    expect(actions.fetchPatientsActivityResponse([1, 2, 3])).toMatchSnapshot()
  })
  describe('fetchPatientsActivity() ', () => {
    // fetch.mockResponse(JSON.stringify({ access_token: '12345' }))
    afterEach(() => {
      fetchMock.restore()
      jest.clearAllMocks()
    })
    it('a successful fetch', () => {
      fetchMock.get('/api/activity', {
        data: [1, 2, 3]
      })
      expect.assertions(3)
      return fetchPatientsActivity(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/activity', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
    it('an unsuccessful fetch', () => {
      fetchMock.get('/api/activity', 402)
      expect.assertions(3)
      return fetchPatientsActivity(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/activity', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
  })
  it('setList', () => {
    expect(actions.setList([1, 2, 3])).toMatchSnapshot()
  })
  it('generatePatientIntensity', () => {
    const patientActivity = [
      {
        activity: 'sleeping',
        minutes: 540
      },
      {
        activity: 'walking',
        minutes: 60
      },
      {
        activity: 'stationary-awake',
        minutes: 770
      },
      {
        activity: 'running',
        minutes: 45
      },
      {
        activity: 'swimming',
        minutes: 45
      }
    ]
    const activities = {
      sleeping: 'none',
      'stationary-awake': 'low',
      walking: 'moderate',
      cycling: 'moderate',
      swimming: 'vigorous',
      running: 'vigorous'
    }
    expect(actions.generatePatientIntensity(patientActivity, activities)).toEqual({
      moderate: 60,
      vigorous: 90
    })
  })
})
