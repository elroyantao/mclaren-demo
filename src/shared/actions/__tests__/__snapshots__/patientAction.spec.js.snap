// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patient actions fetchPatients()  a successful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_PATIENTS_REQUEST",
    },
  ],
  Array [
    Object {
      "patients": Object {
        "data": Array [
          1,
          2,
          3,
        ],
      },
      "type": "FETCH_PATIENTS_RESPONSE",
    },
  ],
]
`;

exports[`patient actions fetchPatients()  an unsuccessful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_PATIENTS_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "FETCH_PATIENTS_FAILURE",
    },
  ],
]
`;

exports[`patient actions fetchPatientsActivity()  a successful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_PATIENTSACTIVITY_REQUEST",
    },
  ],
  Array [
    Object {
      "activities": Object {
        "data": Array [
          1,
          2,
          3,
        ],
      },
      "type": "FETCH_PATIENTSACTIVITY_RESPONSE",
    },
  ],
]
`;

exports[`patient actions fetchPatientsActivity()  an unsuccessful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_PATIENTSACTIVITY_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "FETCH_PATIENTSACTIVITY_FAILURE",
    },
  ],
]
`;

exports[`patient actions fetchPatientsActivityFailure() 1`] = `
Object {
  "type": "FETCH_PATIENTSACTIVITY_FAILURE",
}
`;

exports[`patient actions fetchPatientsActivityRequest() 1`] = `
Object {
  "type": "FETCH_PATIENTSACTIVITY_REQUEST",
}
`;

exports[`patient actions fetchPatientsActivityResponse() 1`] = `
Object {
  "activities": Array [
    1,
    2,
    3,
  ],
  "type": "FETCH_PATIENTSACTIVITY_RESPONSE",
}
`;

exports[`patient actions fetchPatientsFailure() 1`] = `
Object {
  "type": "FETCH_PATIENTS_FAILURE",
}
`;

exports[`patient actions fetchPatientsRequest() 1`] = `
Object {
  "type": "FETCH_PATIENTS_REQUEST",
}
`;

exports[`patient actions fetchPatientsResponse() 1`] = `
Object {
  "patients": Array [
    1,
    2,
    3,
  ],
  "type": "FETCH_PATIENTS_RESPONSE",
}
`;

exports[`patient actions setList 1`] = `
Object {
  "patients": Array [
    1,
    2,
    3,
  ],
  "type": "SET_LIST",
}
`;
